idaes/models/properties/modular_properties/reactions/tests/test_solubility_product_verification.py

```diff
@@ -52,6 +52,7 @@
 
 
 solver = get_solver()
+solver.options["tol"] = 1e-8
```

Comment:
Is this necessary? I am generally hesitant when a test requires changing a solver option as there can be side effects of this.

Reply:
It seems to be necessary for these models for reliable convergence.

I was confused about this at first, as I thought `tol` just affected the convergence criterion. But it turns out it also controls how small the barrier parameter `\mu` can get, based on equation (7) in the [implementation paper](http://cepac.cheme.cmu.edu/pasilectures/biegler/ipopt.pdf).

Occasionally these are converging to a point of local infeasibility with the larger lower bound on `\mu`.

Reply:
So it is probably something that could be fixed with scaling then. For now I think it is probably Ok to do this then.

Reply:
I've been looking at something like this elsewhere, and I may have a suggestion.  I'm having trouble quickly tracing the variables and their bounds, but I speculate that the solid amount and saturation index (or whatever is used here) both have bounds of 0.  The complementarity constraint enforces that one is 0 and the other is more than 0, so the bounds are not necessary.  In my case, when I remove the bounds of the saturation index and solid amount, the problem solves much better.  The problem is hard enough without IPOPT pushing it away from the solution.

Reply:
The problem is having both an active bound for one variable at zero while complementarity simultaneously forces it to be zero. (I think) it causes linearly dependent rows in the KKT system, which causes IPOPT problems and causes dual variables to blow up (because they're no longer unique). I'm afraid of just removing both bounds, though, because then the complementarity might find a solution where both are negative.

Reply:
I'm not sure there is a solution where both are negative.  The constraint is roughly min(SI, Csolid) = 0.  I think that means they both have to be >= 0.